Add Sulu 3.0 compatibility

Prokyonn opened the pull request:
| Q | A
| --- | ---
| Bug fix? | no
| New feature? | no
| BC breaks? | no
| Deprecations? | no
| Fixed tickets | fixes #issuenum
| Related issues/PRs | #issuenum
| License | MIT

#### What's in this PR?

Adds support for Sulu 3.0 

#### To Do

- [ ] Create documentation

Prokyonn committed:
Update dependencies

Prokyonn committed:
Update pool classes to accept iterable for tagged_iterator

Changed FormFieldTypePool and TitleProviderPool constructors to accept
iterable instead of array, enabling Symfony's tagged_iterator injection.

Prokyonn committed:
Remove TaggedServiceCollectorCompilerPass usage

The TaggedServiceCollectorCompilerPass was removed in Sulu 3.0.
Service collection is now handled by Symfony's native tagged_iterator.

Prokyonn committed:
Refactor metadata loaders for Sulu 3.0 API

- PropertiesXmlLoader: Use AdminBundle's AbstractLoader and PropertiesXmlParser
- DynamicFormMetadataLoader: Remove FormMetadataMapper dependency (removed in Sulu 3.0),
  add locales parameter, update to locale-independent caching

Prokyonn committed:
Update service configuration for Sulu 3.0

- Use tagged_iterator for pool services
- Update service references (sulu_page.* -> sulu_admin.*)
- Add locales parameter to DynamicFormMetadataLoader
- Remove FormWebsiteController service registration

Prokyonn committed:
Remove deprecated FormWebsiteController

The controller extended DefaultController which was removed in Sulu 3.0.
Static forms are deprecated and should not be used.

Prokyonn committed:
Replace SingleFormSelection with PropertyResolver and ResourceLoader

- Add SingleFormSelectionPropertyResolver with resourceCallback for entity + serialized data
- Add FormResourceLoader for batch loading forms by ID
- Add FormRepository::loadByIds method
- Add sulu_form_build Twig function for render-time form building
- Remove old SingleFormSelection content type
- Add UPGRADE-3.0.md with migration guide
- Rename UPGRADE.md to UPGRADE-2.x.md

Prokyonn committed:
Refactor for Sulu 3.0: update deprecated Symfony methods and StructureTitleProvider

- Replace isMasterRequest() with isMainRequest() in event listeners
- Replace getMasterRequest() with getMainRequest() in StructureTitleProvider
- Refactor StructureTitleProvider to use DimensionContentInterface instead of removed StructureInterface
- Update UPGRADE-3.0.md with migration notes

Prokyonn committed:
Fix PHPStan errors for Sulu 3.0 compatibility

- Replace ListRepresentation with PaginatedRepresentation
- Fix DateTime vs DateTimeImmutable type errors
- Remove ClassResourceInterface from controllers (removed in FOS REST 3.0)
- Use Request::query->getString() instead of Request::get() with is_string checks
- Fix type annotations and assertions across multiple files
- Regenerate PHPStan baseline

Prokyonn committed:
Update test app routing to use .yaml extension

Prokyonn committed:
Update tests to be compatible with Sulu 3.0

mamazu commented on Tests/Functional/Metadata/DynamicFormMetadataLoaderTest.php:
```diff
@@ -103,15 +103,15 @@ public function testGetMetadataGerman(): void
-        $this->assertEquals('attachment', $fields->getDefaultType());
-        $this->assertEquals([
+        $this->assertEquals('text', $fields->getDefaultType());
+        $this->assertEqualsCanonicalizing([
```
This is a good point, how does 2.6 sort their State fields? Now they're no longer sorted the same way. We could just register the FormTypes with a priority attribute now to force a sorting.

mamazu commented on Metadata/DynamicFormMetadataLoader.php:
```diff
@@ -87,49 +51,49 @@ public function __construct(
+        $types = $this->formFieldTypePool->all();
+
+        $fieldTypeMetaDataCollection = [];
+        foreach ($types as $typeKey => $type) {
+            $a = $this->loadFieldTypeMetadata($typeKey, $type);
+            $fieldTypeMetaDataCollection[$a->getTitle('en')] = $a;
+        }
+        Assert::notEmpty($fieldTypeMetaDataCollection, 'No field type metadata loaded');
```
We're using `addType` that's why we don't need to generate a key explicitly in the array.

```suggestion
        $fieldTypeMetaDataCollection = [];
        foreach ($types as $typeKey => $type) {
            $fieldTypeMetaDataCollection[] = $this->loadFieldTypeMetadata($typeKey, $type);
        }
        Assert::notEmpty($fieldTypeMetaDataCollection, 'No field type metadata loaded');
```

Prokyonn committed:
Update GitHub workflow for Sulu 3.0 compatibility

- Remove PHP 7.2, 7.4, 8.0, 8.1 (incompatible with Sulu 3.0 requirement of PHP ^8.2)
- Add PHP 8.2 with lowest dependencies test
- Add PHP 8.5 with dev stability for future compatibility
- Update MySQL from 5.7 to 8.0
- Update PostgreSQL from 12.5 to 14
- Remove phpunit-9.xml.dist (PHPUnit 10/11 only now)
- Update all database serverVersion parameters

Prokyonn committed:
Add security bundle to website kernel

Prokyonn committed:
Update Metadata/DynamicFormMetadataLoader.php

Co-authored-by: mamazu <14860264+mamazu@users.noreply.github.com>

Prokyonn committed:
Update ramsey composer install

mamazu commented on TitleProvider/TitleProviderPool.php:
```diff
@@ -17,18 +17,16 @@
+    public function __construct(iterable $providers)
     {
-        $this->providers = $providers;
+        $this->providers = $providers instanceof \Traversable ? \iterator_to_array($providers) : $providers;
```
This is how we handle it in the core.

```suggestion
        $this->providers = [...$providers];
```

Prokyonn committed:
Remove usage of CreatePageTrait

alexander-schranz commented:
May some services should be better be internal. But lets merge this and create a first RC release.

alexander-schranz commented:
Thx @Prokyonn @mamazu 👍
